### ISSUE TITLE
2.11.3 consumes CPU

### ISSUE THREAD

**Issue by vlevine** [nats-io/nats.java#482]
Any client, subscriber or publisher, once started, uses CPU almost 100%. 2.11.2 doesn't. Try any of your examples. It was observed on Linux, OpenJDK 11

**Comment by vlevine** [comment]
8 cores. In idle one of cores was 100%.
Attached the sample

[Subscriber.txt](https://github.com/nats-io/nats.java/files/6555194/Subscriber.txt)

**Comment by albertocsm** [comment]
we are also experiencing the same issue.
we see has many 100% cores as (idle) consumers.
still investigating...

**Comment by scottf** [comment]
What version of the server are you running?

**Comment by scottf** [comment]
So I have an 5+ year old laptop that I'm running Ubuntu on. It has 8 cores. I pulled the latest code from github and built it. 
```
./gradlew build -x test
```
Using the cli, I added a stream named `strm` and a subject named `sub` Then I ran these commands, I certainly got a few spikes, but watching with mpstat, the cpu was pretty good:

```
java -cp build/libs/jnats-2.11.3-SNAPSHOT.jar:build/libs/jnats-2.11.3-SNAPSHOT-examples.jar io.nats.examples.jsmulti.JsMulti -a pubAsync -m 1_000_000 -d 3 -u sub
java -cp build/libs/jnats-2.11.3-SNAPSHOT.jar:build/libs/jnats-2.11.3-SNAPSHOT-examples.jar io.nats.examples.jsmulti.JsMulti -a subPush -m 1_000_000 -u sub
```
This is mpstat output on the last 2 runs of subPush
```
Linux 5.8.0-53-generic (batcycle)       05/27/2021      _x86_64_        (8 CPU)

01:32:21 PM  CPU    %usr   %nice    %sys %iowait    %irq   %soft  %steal  %guest  %gnice   %idle
01:32:22 PM  all    0.88    0.00    0.88    0.00    0.00    0.00    0.00    0.00    0.00   98.25
01:32:23 PM  all    0.38    0.00    0.38    0.00    0.00    0.00    0.00    0.00    0.00   99.24
01:32:24 PM  all    4.02    0.00    1.51    0.00    0.00    0.00    0.00    0.00    0.00   94.48
01:32:25 PM  all   51.41    0.00    8.48    0.13    0.00    2.19    0.00    0.00    0.00   37.79
01:32:26 PM  all   35.29    0.00   15.03    0.00    0.00    2.75    0.00    0.00    0.00   46.93
01:32:27 PM  all   35.46    0.00   16.82    0.00    0.00    3.78    0.00    0.00    0.00   43.94
01:32:28 PM  all   30.50    0.00   17.41    0.00    0.00    3.93    0.00    0.00    0.00   48.17
01:32:29 PM  all   31.15    0.00   15.96    0.00    0.00    2.96    0.00    0.00    0.00   49.94
01:32:30 PM  all   30.58    0.00   17.16    0.13    0.00    3.74    0.00    0.00    0.00   48.39
01:32:31 PM  all   31.75    0.00   15.81    0.00    0.00    2.64    0.00    0.00    0.00   49.80
01:32:32 PM  all   30.39    0.00   15.19    0.00    0.00    3.77    0.00    0.00    0.00   50.65
01:32:33 PM  all   44.86    0.00   14.69    0.13    0.00    3.64    0.00    0.00    0.00   36.67
01:32:34 PM  all   41.88    0.00   16.91    0.00    0.00    2.91    0.00    0.00    0.00   38.31
01:32:35 PM  all   42.38    0.00   16.93    0.13    0.00    2.58    0.00    0.00    0.00   37.98
01:32:36 PM  all   40.71    0.00   16.75    0.00    0.00    4.06    0.00    0.00    0.00   38.48
01:32:37 PM  all   49.10    0.00   15.12    0.00    0.00    2.45    0.00    0.00    0.00   33.33
01:32:38 PM  all   40.68    0.00   16.43    0.00    0.00    3.52    0.00    0.00    0.00   39.37
01:32:39 PM  all   43.27    0.00   15.30    0.00    0.00    3.17    0.00    0.00    0.00   38.26
01:32:40 PM  all   42.49    0.00   15.93    0.13    0.00    3.76    0.00    0.00    0.00   37.69
01:32:41 PM  all   42.35    0.00   13.85    0.13    0.00    4.09    0.00    0.00    0.00   39.58
01:32:42 PM  all   29.93    0.00   15.36    0.26    0.00    3.44    0.00    0.00    0.00   50.99
01:32:43 PM  all   12.45    0.00    7.24    0.00    0.00    1.27    0.00    0.00    0.00   79.03
01:32:44 PM  all    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00  100.00
01:32:45 PM  all    0.00    0.00    0.12    0.00    0.00    0.00    0.00    0.00    0.00   99.88
01:32:46 PM  all    0.00    0.00    0.25    0.00    0.00    0.00    0.00    0.00    0.00   99.75
01:32:47 PM  all    0.25    0.00    0.13    0.00    0.00    0.00    0.00    0.00    0.00   99.62
01:32:48 PM  all    0.00    0.00    0.13    0.00    0.00    0.00    0.00    0.00    0.00   99.87
01:32:49 PM  all    0.12    0.00    0.12    0.00    0.00    0.00    0.00    0.00    0.00   99.75
01:32:50 PM  all    0.13    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00   99.87
01:32:51 PM  all   36.50    0.00    6.30    0.13    0.00    0.90    0.00    0.00    0.00   56.17
01:32:52 PM  all   38.98    0.00   13.04    0.00    0.00    3.13    0.00    0.00    0.00   44.85
01:32:53 PM  all   32.11    0.00   14.88    0.00    0.00    3.26    0.00    0.00    0.00   49.74
01:32:54 PM  all   31.02    0.00   14.40    0.00    0.00    3.66    0.00    0.00    0.00   50.92
01:32:55 PM  all   31.36    0.00   14.44    0.26    0.00    3.41    0.00    0.00    0.00   50.52
01:32:56 PM  all   29.80    0.00   15.36    0.00    0.00    3.05    0.00    0.00    0.00   51.79
01:32:57 PM  all   28.90    0.00   16.52    0.00    0.00    3.61    0.00    0.00    0.00   50.97
01:32:58 PM  all   31.23    0.00   16.54    0.00    0.00    3.15    0.00    0.00    0.00   49.08
01:32:59 PM  all   37.43    0.00   15.84    0.00    0.00    3.27    0.00    0.00    0.00   43.46
01:33:00 PM  all   43.42    0.00   14.87    0.13    0.00    3.55    0.00    0.00    0.00   38.03
01:33:01 PM  all   43.21    0.00   15.01    0.00    0.00    3.00    0.00    0.00    0.00   38.77
01:33:02 PM  all   56.20    0.00   13.05    0.00    0.00    3.23    0.00    0.00    0.00   27.52
01:33:03 PM  all   93.99    0.00    5.01    0.00    0.00    0.50    0.00    0.00    0.00    0.50
01:33:04 PM  all   86.11    0.00    4.97    0.00    0.00    0.76    0.00    0.00    0.00    8.15
01:33:05 PM  all   45.25    0.00   13.85    0.26    0.00    3.03    0.00    0.00    0.00   37.60
01:33:06 PM  all   41.53    0.00   16.17    0.00    0.00    3.23    0.00    0.00    0.00   39.07
01:33:07 PM  all   43.54    0.00   16.23    0.00    0.00    3.03    0.00    0.00    0.00   37.20
01:33:08 PM  all   34.03    0.00   16.95    0.00    0.00    3.68    0.00    0.00    0.00   45.34
01:33:09 PM  all   21.92    0.00   13.36    0.00    0.00    3.24    0.00    0.00    0.00   61.48
01:33:10 PM  all    0.38    0.00    0.25    0.00    0.00    0.00    0.00    0.00    0.00   99.37
01:33:11 PM  all    0.00    0.00    0.13    0.00    0.00    0.00    0.00    0.00    0.00   99.87
01:33:12 PM  all    0.25    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00   99.75
01:33:13 PM  all    0.13    0.00    0.00    0.00    0.00    0.00    0.00    0.00    0.00   99.87
Average:     all   28.98    0.00   10.39    0.03    0.00    2.19    0.00    0.00    0.00   58.40
```

**Comment by vlevine** [comment]
I'm quite content with using 2.11.2. Hope it does not affect anyone else.

**Comment by scottf** [comment]
> I'm quite content with using 2.11.2. Hope it does not affect anyone else.

@vlevine Ok so you are working against STAN, not JetStream. And using the the stan.java client and changing it to use the 2.11.3 version of the jnats library? I'm just trying to get context here.

What else can you tell me about your stan setup? Any chance you can move on to JetStream?

**Comment by vlevine** [comment]
current setup is on stan. yes, plan is to move to Jetstream, but not now as we are too close to the prod release.
Using nats streaming.

dependencies from gradle
jnatsVersion is 2.11.3 or 2.11.2
```
    implementation 'io.nats:jnats:' + jnatsVersion
    implementation 'io.nats:java-nats-streaming:2.+'
    implementation 'io.nats:nats-spring:0.4+'
```

**Comment by albertocsm** [comment]
> What version of the server are you running?

nats-streaming:0.8.0-beta

**Comment by albertocsm** [comment]
> > What version of the server are you running?
> 
> nats-streaming:0.8.0-beta

and https://github.com/nats-io/stan.java/blob/2.1.6/

**Comment by lucwillems** [comment]
have a similar experiance, was upgrading my stack (nat server 2.2.6  , java lib from 2.8.0 --> 2.11.3)
using visualVM i captured 2 nats:1 and nats:2 threads eating the CPU away.  8core system with 1 Core running 100%

"nats:1" #21 prio=5 os_prio=0 cpu=83.17ms elapsed=46.38s tid=0x00007f1df831b800 nid=0x570c runnable  [0x00007f1e29d3e000]
   java.lang.Thread.State: RUNNABLE
        at java.net.SocketInputStream.socketRead0(java.base@11.0.11/Native Method)
        at java.net.SocketInputStream.socketRead(java.base@11.0.11/SocketInputStream.java:115)
        at java.net.SocketInputStream.read(java.base@11.0.11/SocketInputStream.java:168)
        at java.net.SocketInputStream.read(java.base@11.0.11/SocketInputStream.java:140)
        at io.nats.client.impl.SocketDataPort.read(SocketDataPort.java:103)
        at io.nats.client.impl.NatsConnectionReader.run(NatsConnectionReader.java:123)
        at java.util.concurrent.Executors$RunnableAdapter.call(java.base@11.0.11/Executors.java:515)
        at java.util.concurrent.FutureTask.run(java.base@11.0.11/FutureTask.java:264)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@11.0.11/ThreadPoolExecutor.java:1128)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@11.0.11/ThreadPoolExecutor.java:628)
        at java.lang.Thread.run(java.base@11.0.11/Thread.java:829)

   Locked ownable synchronizers:
        - <0x0000000629643400> (a java.util.concurrent.ThreadPoolExecutor$Worker)

"nats:2" #22 prio=5 os_prio=0 cpu=45847.46ms elapsed=46.38s tid=0x00007f1df831d000 nid=0x570d runnable  [0x00007f1e29c3c000]
   java.lang.Thread.State: RUNNABLE
        at io.nats.client.impl.NatsConnectionWriter.run(NatsConnectionWriter.java:140)
        at java.util.concurrent.Executors$RunnableAdapter.call(java.base@11.0.11/Executors.java:515)
        at java.util.concurrent.FutureTask.run(java.base@11.0.11/FutureTask.java:264)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@11.0.11/ThreadPoolExecutor.java:1128)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@11.0.11/ThreadPoolExecutor.java:628)
        at java.lang.Thread.run(java.base@11.0.11/Thread.java:829)

seems socket read/write operation keep eating cpu while processing 10 messages / sec (smal test message...)

going to 2.11.2 , seem ok , no  CPU burning an steady 10msg/sec processing .

**Comment by lucwillems** [comment]
suspect that based on cpu=45847ms with elapse time 46.38s , that nats:2 is eating the most cpu ...

locking at the code, its seems never to wait a bit and has a continuous loop in run() method , sometimes jumps in the  
if (mcount>0) {}  block, but most of the time just racing in the loop.

**Comment by scottf** [comment]
2.11.3 is definitely bad. I'm working on the fix.

**Comment by scottf** [comment]
I have released 2.11.4, reverting back to code in 2.11.2. Thanks for reporting this and sorry for this problem. Going to take another pass at improving, hopefully will have a good release soon.
